websocket-ballerina/Package.md

```diff
@@ -1,32 +1,33 @@
 ## Module Overview
 
-This module provides an implementation for connecting and interacting with WebSocket endpoints. The module facilitates two types of network entry points as ‘Client’ and ‘Listener’.
+This module provides an implementation for connecting and interacting with WebSocket endpoints. The module facilitates three types of network entry points as ‘AsyncClient’, ‘SyncClient’ and ‘Listener’.
 
-### WebSocket Module
+The `AsyncClient` has a callback service that can be registered at the initialization of the client. It has a fixed set of remote methods in this service and they get called on the receipt of messages from the server. The ‘SyncClient’ also can have a callback service registered to receive control frames like ping, pong and close.
 
-This module also provides support for WebSockets. There are two types of WebSocket endpoints: `WebSocketClient` and `WebSocketListener`. Both endpoints support all WebSocket frames. The `WebSocketClient` has a callback service.
+In the server side, initial websocket service is there to handle upgrade requests. It has a single `onUpgrade` resource which takes in `http:Caller` and `http:Request` optionally. The `onUpgrade` resource returns a `websocket:Service` to which incoming messages gets dispatched after a successful websocket connection upgrade. This resource can be used to intercept the initial HTTP upgrade with custom headers or to cancel the websocket upgrade by returning an error.
+The returning `websocket:Service` has a fixed set of remote methods.
 
-There are two types of services for WebSockets. The service of the server has the `WebSocketCaller` as the resource parameter and the callback service of the client has `WebSocketClient` as the resource parameter. The WebSocket services have a fixed set of resources that do not have a resource config. The incoming messages are passed to these resources.
-
-**WebSocket upgrade**: During a WebSocket upgrade, the initial message received is an HTTP request. To intercept this request and perform the upgrade explicitly with custom headers, the user must create an HTTP resource with WebSocket-specific configurations as follows:
+**WebSocket upgrade**: During a WebSocket upgrade, the initial message received is an HTTP request. 
 
 ```ballerina
-@http:ResourceConfig {
-    webSocketUpgrade: {
-        upgradePath: "/{name}",
-        upgradeService: chatApp
-    }
-}
-resource function upgrader(http:Caller caller, http:Request req, string name) {
+service /basePath on new websocket:Listener(21003) {
+    resource function onUpgrade .(http:Caller caller, http:Request req) returns websocket:Service|websocket:UpgradeError {
```

Comment:
Is this correct ? I thought it should be 

`
resource function get .(http:Caller caller, http:Request req) returns websocket:Service|websocket:UpgradeError
`

Reply:
It is the resource(method) name, right? In HTTP it is mapped to the accessor get, post, etc. In WebSocket, we always get a specific get request which is always dispatched to this single resource. I think it is clear to have `onUpgrade` as the name in websocket context.

Reply:
Actually what comes first is the accessor name. @chamil321 can you confirm this ?

Reply:
Yes, The first identifier is the accessor name.  The second one is the resource name(path). In this case she has used `onUpgrade` as accessor name which is a custom accessor. Here the HTTP dispatching logic does not apply as webscoket as its own implementation.

Reply:
IMO, this could lead to confusions as we maintain all the HTTP semantics in this resource. Initial request is HTTP and HTTP does not have an onUpgrade verb. 

@Bhashinee it is not the resource name. So what we need is

resource function get ws(http:Caller caller, http:Request req) returns websocket:Service|websocket:UpgradeError